processor/stream/stream_processor.go

```diff
@@ -0,0 +1,239 @@
+	if err != nil {
+		return nil, &Error{
+			Type:     InvalidInputErrType,
+			Message:  err.Error(),
+			Document: string(reader.LastLine()),
+		}
+	}
+
+	// create a metadata struct
+	metadata, err := metadata.DecodeMetadata(rawMetadata)
+	if err != nil {
+		return nil, err
+	}
+
+	return metadata, nil
+}
+
+// handleRawModel validates and decodes a single json object into its struct form
+func (v *StreamProcessor) handleRawModel(rawModel map[string]interface{}) (transform.Transformable, error) {
+	for _, model := range models {
+		if entry, ok := rawModel[model.key]; ok {
+			err := validation.Validate(entry, model.schema)
+			if err != nil {
+				return nil, err
+			}
+
+			tr, err := model.modelDecoder(entry, err)
+			if err != nil {
+				return nil, err
+			}
+			return tr, nil
+		}
+	}
+	return nil, ErrUnrecognizedObject
+}
+
+// readBatch will read up to `batchSize` objects from the ndjson stream
+// it returns a slice of eventables, a serverResponse and a bool that indicates if we're at EOF.
+func (v *StreamProcessor) readBatch(batchSize int, reader StreamReader, response *Result) ([]transform.Transformable, bool) {
+	var err error
+	var rawModel map[string]interface{}
+
+	var eventables []transform.Transformable
+	for i := 0; i < batchSize && err == nil; i++ {
+		rawModel, err = reader.Read()
+		if err != nil && err != io.EOF {
+
+			if e, ok := err.(decoder.JSONDecodeError); ok {
+				response.LimitedAdd(&Error{
+					Type:     InvalidInputErrType,
+					Message:  e.Error(),
+					Document: string(reader.LastLine()),
+				})
+				continue
+			}
+
+			// return early, we assume we can only recover from a JSON decode error
+			response.LimitedAdd(err)
+			return eventables, true
+		}
+
+		if rawModel != nil {
+			tr, err := v.handleRawModel(rawModel)
+			if err != nil {
+				response.LimitedAdd(&Error{
+					Type:     InvalidInputErrType,
+					Message:  err.Error(),
+					Document: string(reader.LastLine()),
+				})
+				continue
+			}
+			eventables = append(eventables, tr)
+		}
+	}
+
+	return eventables, reader.IsEOF()
+}
+
+func (s *StreamProcessor) HandleStream(ctx context.Context, meta map[string]interface{}, jsonReader StreamReader, report publish.Reporter) *Result {
+	res := &Result{}
+	metadata, err := s.readMetadata(meta, jsonReader)
+	// no point in continuing if we couldn't read the metadata
+	if err != nil {
+		res.LimitedAdd(err)
+		return res
+	}
+
+	tctx := &transform.Context{
+		RequestTime: utility.RequestTime(ctx),
+		Config:      s.Tconfig,
+		Metadata:    *metadata,
+	}
+
+	for {
+		transformables, done := s.readBatch(batchSize, jsonReader, res)
+		if transformables != nil {
+			err := report(ctx, publish.PendingReq{
+				Transformables: transformables,
+				Tcontext:       tctx,
+			})
+
+			if err != nil {
+				switch err {
+				case publish.ErrChannelClosed:
+					res.LimitedAdd(&Error{
+						Type:    ShuttingDownErrType,
+						Message: "server is shutting down",
+					})
+				case publish.ErrFull:
```

Comment:
As https://github.com/roncohen/apm-server/pull/3 has not been merged, I guess this will need to be implemented again after merging this PR.

Reply:
I needed to refactor this area, so your PR wouldn't apply cleanly anymore. During the refactor i changed the behavior in this PR so that it returns when the queue is full, like in your PR.